helper/logging/logging_http_transport.go

```diff
@@ -0,0 +1,247 @@
+	FieldHttpRequestProtoVersion = "tf_http_req_version"
+
+	// OperationHttpResponse is the field value used by NewSubsystemLoggingHTTPTransport when logging a response via tflog.
+	OperationHttpResponse = "response"
+
+	// FieldHttpResponseProtoVersion is the field key used by NewSubsystemLoggingHTTPTransport when logging a response HTTP protocol version via tflog.
+	FieldHttpResponseProtoVersion = "tf_http_res_version"
+
+	// FieldHttpResponseStatusCode is the field key used by NewSubsystemLoggingHTTPTransport when logging a response status code via tflog.
+	FieldHttpResponseStatusCode = "tf_http_res_status_code"
+
+	// FieldHttpResponseStatusReason is the field key used by NewSubsystemLoggingHTTPTransport when logging a response status reason phrase via tflog.
+	FieldHttpResponseStatusReason = "tf_http_res_status_reason"
+)
+
+type loggingHttpTransport struct {
+	subsystem      string
+	logToSubsystem bool
+	transport      http.RoundTripper
+}
+
+func (t *loggingHttpTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	ctx := req.Context()
+
+	//// Decompose the request bytes in a message (HTTP body) and fields (HTTP headers), then log it
+	msg, fields, err := decomposeRequestForLogging(req)
+	if err != nil {
+		t.Error(ctx, "Failed to parse request bytes for logging", map[string]interface{}{
+			"error": err,
+		})
+	} else {
+		t.Debug(ctx, msg, fields)
+	}
+
+	// Invoke the wrapped RoundTrip now
+	res, err := t.transport.RoundTrip(req)
+	if err != nil {
+		return res, err
+	}
+
+	// Decompose the response bytes in a message (HTTP body) and fields (HTTP headers), then log it
+	msg, fields, err = decomposeResponseForLogging(res)
+	if err != nil {
+		t.Error(ctx, "Failed to parse response bytes for logging", map[string]interface{}{
+			"error": err,
+		})
+	} else {
+		t.Debug(ctx, msg, fields)
+	}
+
+	return res, nil
+}
+
+func (t *loggingHttpTransport) Debug(ctx context.Context, msg string, fields ...map[string]interface{}) {
+	if t.logToSubsystem {
+		tflog.SubsystemDebug(ctx, t.subsystem, msg, fields...)
+	} else {
+		tflog.Debug(ctx, msg, fields...)
+	}
+}
+
+func (t *loggingHttpTransport) Error(ctx context.Context, msg string, fields ...map[string]interface{}) {
+	if t.logToSubsystem {
+		tflog.SubsystemError(ctx, t.subsystem, msg, fields...)
+	} else {
+		tflog.Error(ctx, msg, fields...)
+	}
+}
+
+func decomposeRequestForLogging(req *http.Request) (string, map[string]interface{}, error) {
+	fields := make(map[string]interface{}, len(req.Header)+4)
+	fields[FieldHttpOperationType] = OperationHttpRequest
+
+	fields[FieldHttpRequestMethod] = req.Method
+	fields[FieldHttpRequestUri] = req.URL.RequestURI()
+	fields[FieldHttpRequestProtoVersion] = req.Proto
+
+	// Get the full body of the request, including headers appended by http.Transport:
+	// this is necessary because the http.Request at this stage doesn't contain
+	// all the headers that will be eventually sent.
```

Comment:
It's a bummer that these headers go beyond just `User-Agent` (which maybe we should set one to mention `terraform-plugin-sdk`, if its not already set?), since that would certainly simplify this logic.

Reply:
Yeah, that's why this `http.DumpRequestOut` utility is so tricky, because internally it essentially "pretends" to send the request, and grabs the output at the "end of the pipe" (if it makes sense).

Reply:
But at least you made me get rid of manual parsing and relying on `textproto`, and that's good.